File: video-parse/src/handler.rs
use axum::{
    extract::Json,
    http::StatusCode,
    response::Json as ResponseJson,
};
use std::path::PathBuf;
use crate::{OssEvent, ProcessResponse, ProcessResult, OssClient, ProcessConfig, process_video};
use tracing::{info, error};

/// 处理 OSS Event 的 Handler
pub async fn handle_oss_event(
    Json(event): Json<OssEvent>,
) -> Result<ResponseJson<ProcessResponse>, (StatusCode, String)> {
    info!("收到 OSS Event: {:?}", event);

    // DEBUG 模式：如果设置了 DEBUG=true，直接返回成功，用于测试部署和事件触发
    if std::env::var("DEBUG")
        .unwrap_or_else(|_| "false".to_string())
        .to_lowercase() == "true"
    {
        info!("DEBUG 模式已启用，跳过实际处理，直接返回成功");
        
        // 提取事件信息用于日志
        let event_info = if !event.events.is_empty() {
            let event_item = &event.events[0];
            format!(
                "bucket={}, key={}, region={}, eventName={}",
                event_item.oss.bucket.name,
                event_item.oss.object.key,
                event_item.region,
                event_item.event_name
            )
        } else {
            "无事件信息".to_string()
        };
        
        info!("DEBUG 模式 - 事件信息: {}", event_info);
        
        return Ok(ResponseJson(ProcessResponse {
            success: true,
            message: format!("DEBUG 模式：事件接收成功，事件信息: {}", event_info),
            result: None,
        }));
    }

    // 检查是否有事件
    if event.events.is_empty() {
        return Err((
            StatusCode::BAD_REQUEST,
            "事件列表为空".to_string(),
        ));
    }

    // 处理第一个事件（通常只有一个）
    let event_item = &event.events[0];
    
    // 只处理 ObjectCreated 事件（包括 Put, Post, Copy, CompleteMultipartUpload, PutSymlink）
    if !event_item.event_name.starts_with("ObjectCreated") {
        return Ok(ResponseJson(ProcessResponse {
            success: false,
            message: format!("不支持的事件类型: {}", event_item.event_name),
            result: None,
        }));
    }

    let bucket = &event_item.oss.bucket.name;
    let mut object_key = event_item.oss.object.key.clone();
    let region = &event_item.region;
    
    // 处理符号链接事件（参考 Python 示例）
    if event_item.event_name == "ObjectCreated:PutSymlink" {
        // TODO: 实现符号链接解析
        // 在函数计算环境中，符号链接需要通过 OSS API 解析
        // 当前先记录日志，后续可以实现 get_symlink 功能
        info!("检测到符号链接事件，需要解析符号链接: {}", object_key);
        // 注意：符号链接解析需要 OSS SDK 支持，当前暂不处理
        return Ok(ResponseJson(ProcessResponse {
            success: false,
            message: format!("暂不支持符号链接事件: {}", event_item.event_name),
            result: None,
        }));
    }
    
    info!("处理视频文件: bucket={}, key={}, region={}", bucket, object_key, region);

    // 创建临时目录
    // 尝试使用函数计算的 request_id（如果可用），否则使用时间戳+UUID
    let request_id = std::env::var("FC_REQUEST_ID")
        .unwrap_or_else(|_| {
            format!("{}_{}", 
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_secs(),
                uuid::Uuid::new_v4().to_string()
            )
        });
    let temp_dir = std::env::temp_dir().join("video-parse").join(&request_id);
    std::fs::create_dir_all(&temp_dir)
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("创建临时目录失败: {}", e)))?;

    // 下载视频文件
    let oss_client = OssClient::new()
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, format!("创建 OSS 客户端失败: {}", e)))?;
    
    let video_filename = PathBuf::from(&object_key)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("video.mp4");
    
    let video_path = temp_dir.join(video_filename);
    
    // 构建 internal endpoint（内网访问更快且免费）
    // 格式: oss-{region}-internal.aliyuncs.com
    let endpoint = format!("oss-{}-internal.aliyuncs.com", region);
    
    // 尝试从 OSS 下载（使用 internal endpoint）
    let downloaded_path = oss_client
        .download_file(bucket, &object_key, Some(&endpoint), &video_path)
        .await
        .map_err(|e| {
            error!("下载文件失败: {}", e);
            (StatusCode::INTERNAL_SERVER_ERROR, format!("下载文件失败: {}", e))
        })?;

    // 创建输出目录
    let output_dir = temp_dir.join("output");
    
    // 处理视频
    let config = ProcessConfig::default();
    let process_result = process_video(&downloaded_path, &output_dir, config)
        .await
        .map_err(|e| {
            error!("处理视频失败: {}", e);
            (StatusCode::INTERNAL_SERVER_ERROR, format!("处理视频失败: {}", e))
        })?;

    // 上传处理结果到目标 bucket（如果配置了目标 bucket）
    let uploaded_files = if let (Some(dest_bucket), Some(dest_region)) = (
        std::env::var("DESTINATION_BUCKET").ok(),
        std::env::var("DESTINATION_REGION").ok(),
    ) {
        info!("开始上传处理结果到目标 bucket: {}/{}", dest_region, dest_bucket);
        
        // 构建目标 endpoint
        let dest_endpoint = format!("oss-{}-internal.aliyuncs.com", dest_region);
        
        // 构建目标路径前缀（保持源文件的目录结构）
        let dest_prefix = std::env::var("DESTINATION_PREFIX")
            .unwrap_or_else(|_| {
                // 默认使用源文件的目录部分作为前缀
                PathBuf::from(&object_key)
                    .parent()
                    .and_then(|p| p.to_str())
                    .map(|s| s.to_string())
                    .unwrap_or_else(|| "processed".to_string())
            });
        
        let mut uploaded = Vec::new();
        let mut upload_errors = Vec::new();
        
        // 上传关键帧
        // keyframe_files 是文件名列表，需要与 output_dir 组合成完整路径
        for keyframe_filename in &process_result.keyframe_files {
            let keyframe_path = output_dir.join(keyframe_filename);
            if keyframe_path.exists() {
                let keyframe_name = keyframe_path.file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or("keyframe.jpg");
                let dest_key = format!("{}/keyframes/{}", dest_prefix, keyframe_name);
                
                match oss_client.upload_file(
                    &dest_bucket,
                    &dest_key,
                    &keyframe_path,
                    Some(&dest_endpoint),
                ).await {
                    Ok(_) => {
                        info!("已上传关键帧: {} -> {}", keyframe_path.display(), dest_key);
                        uploaded.push(dest_key.clone());
                    }
                    Err(e) => {
                        let error_msg = format!("上传关键帧失败 {}: {}", dest_key, e);
                        error!("{}", error_msg);
                        upload_errors.push(error_msg);
                    }
                }
            } else {
                let error_msg = format!("关键帧文件不存在: {}", keyframe_path.display());
                error!("{}", error_msg);
                upload_errors.push(error_msg);
            }
        }
        
        // 上传音频文件
        // audio_file 是文件名，需要与 output_dir 组合成完整路径
        let audio_path = output_dir.join(&process_result.audio_file);
        if audio_path.exists() {
            let audio_name = audio_path.file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("audio.aac");
            let dest_key = format!("{}/{}", dest_prefix, audio_name);
            
            match oss_client.upload_file(
                &dest_bucket,
                &dest_key,
                &audio_path,
                Some(&dest_endpoint),
            ).await {
                Ok(_) => {
                    info!("已上传音频文件: {} -> {}", audio_path.display(), dest_key);
                    uploaded.push(dest_key.clone());
                }
                Err(e) => {
                    let error_msg = format!("上传音频文件失败 {}: {}", dest_key, e);
                    error!("{}", error_msg);
                    upload_errors.push(error_msg);
                }
            }
        } else {
            let error_msg = format!("音频文件不存在: {}", audio_path.display());
            error!("{}", error_msg);
            upload_errors.push(error_msg);
        }
        
        // 上传元数据文件
        let metadata_path = output_dir.join("metadata.json");
        if metadata_path.exists() {
            let dest_key = format!("{}/metadata.json", dest_prefix);
            
            match oss_client.upload_file(
                &dest_bucket,
                &dest_key,
                &metadata_path,
                Some(&dest_endpoint),
            ).await {
                Ok(_) => {
                    info!("已上传元数据文件: {} -> {}", metadata_path.display(), dest_key);
                    uploaded.push(dest_key.clone());
                }
                Err(e) => {
                    let error_msg = format!("上传元数据文件失败 {}: {}", dest_key, e);
                    error!("{}", error_msg);
                    upload_errors.push(error_msg);
                }
            }
        } else {
            let error_msg = format!("元数据文件不存在: {}", metadata_path.display());
            error!("{}", error_msg);
            upload_errors.push(error_msg);
        }
        
        // 记录上传结果
        if !upload_errors.is_empty() {
            error!("部分文件上传失败，共 {} 个错误", upload_errors.len());
            for err in &upload_errors {
                error!("  - {}", err);
            }
        }
        
        info!("上传完成: 成功 {} 个，失败 {} 个", uploaded.len(), upload_errors.len());
        
        Some(uploaded)
    } else {
        info!("未配置目标 bucket，跳过上传");
        None
    };

    // 清理临时目录（可选，函数计算会自动清理）
    // 如果需要保留文件用于调试，可以注释掉下面的代码
    // if let Err(e) = std::fs::remove_dir_all(&temp_dir) {
    //     error!("清理临时目录失败: {}", e);
    // }

    // 构建响应
    let response = ProcessResponse {
        success: true,
        message: format!(
            "成功处理视频，检测到 {} 个场景{}",
            process_result.metadata.scene_count,
            if uploaded_files.is_some() {
                "，已上传到目标 bucket"
            } else {
                ""
            }
        ),
        result: Some(ProcessResult {
            video_file: downloaded_path.to_string_lossy().to_string(),
            output_dir: output_dir.to_string_lossy().to_string(),
            scene_count: process_result.metadata.scene_count,
            keyframes: process_result.keyframe_files,
            audio_file: process_result.audio_file,
            metadata_file: "metadata.json".to_string(),
        }),
    };

    info!("处理完成: {:?}", response);

    Ok(ResponseJson(response))
}

/// 健康检查 Handler
pub async fn health_check() -> &'static str {
    "OK"
}

File: code/handler/handler.go
package handler

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/google/uuid"
	"velocn-video-parse/config"
	"velocn-video-parse/oss"
	"velocn-video-parse/types"
	"velocn-video-parse/video"
)

// Handler HTTP 请求处理器
type Handler struct {
	configLoader *config.ConfigLoader
}

// NewHandler 创建新的处理器
func NewHandler() *Handler {
	return &Handler{
		configLoader: &config.ConfigLoader{},
	}
}

// isVideoFile 检查文件是否是视频文件（通过扩展名）
func isVideoFile(filename string) bool {
	ext := strings.ToLower(filepath.Ext(filename))
	videoExtensions := []string{
		".mp4", ".avi", ".mov", ".mkv", ".flv", ".wmv", ".webm",
		".m4v", ".3gp", ".ogv", ".ts", ".mts", ".m2ts",
		".vob", ".asf", ".rm", ".rmvb", ".divx", ".xvid",
	}
	for _, videoExt := range videoExtensions {
		if ext == videoExt {
			return true
		}
	}
	return false
}

// HandleRequest 处理 HTTP 请求
func (h *Handler) HandleRequest(method, path string, body []byte, headers map[string]string, queryParams map[string]string) (*types.HTTPTriggerResponse, error) {
	// 添加 CORS 头
	corsHeaders := map[string]string{
		"Access-Control-Allow-Origin":  "*",
		"Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, PATCH, HEAD, OPTIONS",
		"Access-Control-Allow-Headers": "Content-Type, Authorization, x-fc-request-id",
	}

	// 处理 CORS 预检请求
	if method == "OPTIONS" {
		return types.NewHTTPTriggerResponse(200).WithHeaders(corsHeaders), nil
	}

	// 路由处理
	switch {
	case path == "/" || path == "/health":
		return h.HandleHealthCheck(corsHeaders), nil
	case path == "/initialize":
		return h.handleInitialize(corsHeaders), nil
	case strings.HasPrefix(path, "/invoke"):
		return h.handleInvoke(method, body, corsHeaders), nil
	case strings.HasPrefix(path, "/process"):
		if path == "/process/direct" {
			return h.handleDirectProcess(body, corsHeaders), nil
		} else if path == "/process/query" {
			return h.handleProcessQuery(queryParams, corsHeaders), nil
		} else {
			return h.handleOSSEventRequest(body, corsHeaders), nil
		}
	default:
		return types.NewHTTPTriggerResponse(404).
			WithHeaders(corsHeaders).
			WithBody(fmt.Sprintf(`{"success":false,"message":"未找到路径: %s"}`, path)), nil
	}
}

// HandleHealthCheck 健康检查（公开方法）
func (h *Handler) HandleHealthCheck(corsHeaders map[string]string) *types.HTTPTriggerResponse {
	response := map[string]interface{}{
		"success": true,
		"message": "server is running",
		"data": map[string]interface{}{
			"status":    "ok",
			"timestamp": time.Now().Format(time.RFC3339),
		},
	}
	body, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(body))
}

// handleInitialize 函数计算初始化端点
func (h *Handler) handleInitialize(corsHeaders map[string]string) *types.HTTPTriggerResponse {
	response := map[string]interface{}{
		"success": true,
		"message": "FunctionCompute 初始化完成",
		"data": map[string]interface{}{
			"timestamp": time.Now().Format(time.RFC3339),
		},
	}
	body, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(body))
}

// handleInvoke 函数计算调用端点
func (h *Handler) handleInvoke(method string, body []byte, corsHeaders map[string]string) *types.HTTPTriggerResponse {
	// 尝试解析为 OSS 事件
	if len(body) > 0 {
		var ossEvent oss.OssEvent
		if err := json.Unmarshal(body, &ossEvent); err == nil && len(ossEvent.Events) > 0 {
			log.Printf("检测到 OSS 事件，开始处理...")
			// 异步处理
			go func() {
				_, _ = h.HandleOSSEvent(&ossEvent)
			}()
		}
	}

	response := map[string]interface{}{
		"success": true,
		"message": "请求已接收",
		"data": map[string]interface{}{
			"timestamp": time.Now().Format(time.RFC3339),
		},
	}
	responseBody, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(responseBody))
}

// handleOSSEventRequest 处理 OSS 事件请求
func (h *Handler) handleOSSEventRequest(body []byte, corsHeaders map[string]string) *types.HTTPTriggerResponse {
	if len(body) == 0 {
		response := types.ProcessResponse{
			Success: false,
			Message: "请求体为空",
		}
		responseBody, _ := json.Marshal(response)
		return types.NewHTTPTriggerResponse(400).
			WithHeaders(corsHeaders).
			WithBody(string(responseBody))
	}

	var ossEvent oss.OssEvent
	if err := json.Unmarshal(body, &ossEvent); err != nil {
		response := types.ProcessResponse{
			Success: false,
			Message: fmt.Sprintf("解析 JSON 失败: %v", err),
		}
		responseBody, _ := json.Marshal(response)
		return types.NewHTTPTriggerResponse(400).
			WithHeaders(corsHeaders).
			WithBody(string(responseBody))
	}

	// 处理 OSS 事件
	response, err := h.HandleOSSEvent(&ossEvent)
	if err != nil {
		errorResponse := types.ProcessResponse{
			Success: false,
			Message: err.Error(),
		}
		responseBody, _ := json.Marshal(errorResponse)
		return types.NewHTTPTriggerResponse(500).
			WithHeaders(corsHeaders).
			WithBody(string(responseBody))
	}

	responseBody, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(responseBody))
}

// HandleOSSEvent 处理 OSS 事件
func (h *Handler) HandleOSSEvent(event *oss.OssEvent) (*types.ProcessResponse, error) {
	log.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	log.Printf("[OSS Event] Received OSS event trigger request")
	log.Printf("Event count: %d", len(event.Events))

	if len(event.Events) == 0 {
		return &types.ProcessResponse{
			Success: false,
			Message: "事件列表为空",
		}, nil
	}

	eventItem := event.Events[0]

	if !strings.HasPrefix(eventItem.EventName, "ObjectCreated") {
		log.Printf("[OSS Event] Unsupported event type: %s, skipping processing", eventItem.EventName)
		return &types.ProcessResponse{
			Success: false,
			Message: fmt.Sprintf("不支持的事件类型: %s", eventItem.EventName),
		}, nil
	}

	// 加载扩展配置
	extendedConfig := h.configLoader.LoadExtendedConfig("")

	if extendedConfig.DebugMode {
		log.Printf("DEBUG mode enabled, skipping actual processing")
		return &types.ProcessResponse{
			Success: true,
			Message: "DEBUG 模式：事件接收成功",
		}, nil
	}

	bucket := eventItem.Oss.Bucket.Name
	objectKey := eventItem.Oss.Object.Key
	region := eventItem.Region

	// 检查文件类型，只处理视频文件
	if !isVideoFile(objectKey) {
		log.Printf("[OSS Event] 跳过非视频文件: %s", objectKey)
		return &types.ProcessResponse{
			Success: true,
			Message: fmt.Sprintf("跳过非视频文件: %s", objectKey),
		}, nil
	}

	log.Printf("[OSS Event] 检测到视频文件，开始处理: %s", objectKey)

	// 创建临时目录
	requestID := os.Getenv("FC_REQUEST_ID")
	if requestID == "" {
		requestID = fmt.Sprintf("%d_%s", time.Now().Unix(), uuid.New().String())
	}

	tempDir := ""
	if extendedConfig.OutputPath != "" {
		tempDir = filepath.Join(extendedConfig.OutputPath, requestID)
	} else {
		tempDir = filepath.Join(os.TempDir(), "video-parse", requestID)
	}

	if err := os.MkdirAll(tempDir, 0755); err != nil {
		return nil, fmt.Errorf("创建临时目录失败: %v", err)
	}

	// 步骤 1: 下载视频文件
	log.Printf("📥 [OSS Event] 步骤 1/3: 开始下载视频文件...")
	log.Printf("  • Bucket: %s", bucket)
	log.Printf("  • Object Key: %s", objectKey)
	log.Printf("  • Region: %s", region)
	log.Printf("  • 文件大小: %d 字节", eventItem.Oss.Object.Size)

	ossClient, err := oss.NewOssClient()
	if err != nil {
		return nil, fmt.Errorf("创建 OSS 客户端失败: %v", err)
	}

	videoPathBuf := filepath.Base(objectKey)
	videoFilename := videoPathBuf
	if videoFilename == "" || videoFilename == "." {
		videoFilename = "video.mp4"
	}

	videoPath := filepath.Join(tempDir, videoFilename)
	endpoint := fmt.Sprintf("oss-%s-internal.aliyuncs.com", region)

	log.Printf("  • 下载到本地路径: %s", videoPath)
	downloadedPath, err := ossClient.DownloadFile(bucket, objectKey, endpoint, videoPath)
	if err != nil {
		return nil, fmt.Errorf("下载文件失败: %v", err)
	}
	log.Printf("✅ [OSS Event] 步骤 1/3: 视频文件下载完成")
	log.Printf("  • 本地文件路径: %s", downloadedPath)

	// 步骤 2: 创建输出目录
	log.Printf("📁 [OSS Event] 步骤 2/3: 准备输出目录...")
	outputDir := filepath.Join(tempDir, "output")
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, fmt.Errorf("创建输出目录失败: %v", err)
	}
	log.Printf("✅ [OSS Event] 步骤 2/3: 输出目录准备完成")
	log.Printf("  • 输出目录: %s", outputDir)

	// 步骤 3: 处理视频（拉片）
	log.Printf("🎬 [OSS Event] 步骤 3/3: 开始处理视频（拉片）...")
	log.Printf("  • 处理配置:")
	log.Printf("    - 阈值: %.2f", extendedConfig.Process.Threshold)
	log.Printf("    - 最小场景时长: %.2f 秒", extendedConfig.Process.MinSceneDuration)
	log.Printf("    - 采样率: %.2f fps", extendedConfig.Process.SampleRate)

	cfg := extendedConfig.Process
	processResult, err := video.ProcessVideo(downloadedPath, outputDir, cfg)
	if err != nil {
		return nil, fmt.Errorf("处理视频失败: %v", err)
	}

	log.Printf("✅ [OSS Event] 步骤 3/3: 视频处理完成")
	log.Printf("  • 检测到场景数: %d 个", processResult.Metadata.SceneCount)
	log.Printf("  • 提取关键帧数: %d 个", len(processResult.KeyframeFiles))
	if processResult.AudioFile != "" {
		log.Printf("  • 音频文件: %s", processResult.AudioFile)
	}

	// 构建响应
	log.Printf("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	log.Printf("✅ [OSS Event] 所有处理步骤完成")
	return &types.ProcessResponse{
		Success: true,
		Message: fmt.Sprintf("成功处理视频，检测到 %d 个场景，提取 %d 个关键帧",
			processResult.Metadata.SceneCount,
			len(processResult.KeyframeFiles)),
		Result: &types.ProcessResult{
			VideoFile:    downloadedPath,
			OutputDir:    outputDir,
			SceneCount:   processResult.Metadata.SceneCount,
			Keyframes:    processResult.KeyframeFiles,
			AudioFile:    processResult.AudioFile,
			MetadataFile: "metadata.json",
		},
	}, nil
}

// handleDirectProcess 直接处理请求
func (h *Handler) handleDirectProcess(body []byte, corsHeaders map[string]string) *types.HTTPTriggerResponse {
	var request struct {
		Input            string   `json:"input"`
		Output           string   `json:"output"`
		Threshold        *float64 `json:"threshold"`
		MinSceneDuration *float64 `json:"min_scene_duration"`
		SampleRate       *float64 `json:"sample_rate"`
		IsOSSPath        *bool    `json:"is_oss_path"`
		OSSBucket        string   `json:"oss_bucket"`
		OSSRegion        string   `json:"oss_region"`
	}

	if err := json.Unmarshal(body, &request); err != nil {
		response := types.ProcessResponse{
			Success: false,
			Message: fmt.Sprintf("解析 JSON 失败: %v", err),
		}
		responseBody, _ := json.Marshal(response)
		return types.NewHTTPTriggerResponse(400).
			WithHeaders(corsHeaders).
			WithBody(string(responseBody))
	}

	// 这里简化处理，实际应该调用完整的处理逻辑
	response := types.ProcessResponse{
		Success: true,
		Message: "处理请求已接收",
	}
	responseBody, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(responseBody))
}

// handleProcessQuery 通过查询参数处理视频
func (h *Handler) handleProcessQuery(queryParams map[string]string, corsHeaders map[string]string) *types.HTTPTriggerResponse {
	input := queryParams["input"]
	if input == "" {
		response := types.ProcessResponse{
			Success: false,
			Message: "缺少 input 参数",
		}
		responseBody, _ := json.Marshal(response)
		return types.NewHTTPTriggerResponse(400).
			WithHeaders(corsHeaders).
			WithBody(string(responseBody))
	}

	// 这里简化处理，实际应该调用完整的处理逻辑
	response := types.ProcessResponse{
		Success: true,
		Message: "处理请求已接收",
	}
	responseBody, _ := json.Marshal(response)
	return types.NewHTTPTriggerResponse(200).
		WithHeaders(corsHeaders).
		WithBody(string(responseBody))
}
